[DO NOT MERGE] Add WDQMS processing to GFSv16.3.x

aerorahul opened the pull request:
**This PR will be merged in a release branch.  This PR is only for review and testing.**

# Description
This PR:
- brings in wdqms processing code from https://github.com/kevindougherty-noaa/wdqms into the global-workflow as EE2 does not allow codes in personal repos and combines the driver with the tool.
- adds the rocoto job, j-job, exscript, config file and updates the configs and resources, etc. for the additional `wdqms` job.
- adds a task in the rocoto and ecflow suites

TODO:
- Identify `COMIN` and `COMOUT` for the data going into and coming out of`wdqms.py`
- Identify modules in the `ecf` that need to be loaded on wcoss2
- Identify modules on dev machines (if not already done via `load_fv3gfs_modules.sh`

Fixes #1459 

# Type of change
<!-- Delete all except one -->
- New feature (adds functionality)

# Change characteristics
- Is this a breaking change (a change in existing functionality)? NO
- Does this change require a documentation update? YES

# How has this been tested?
<!-- Please list any test you conducted, including the machine.

Example:
- Clone and build on WCOSS
- Cycled test on Orion
- Forecast-only on Hera
-->

# Checklist
- [ ] Any dependent changes have been merged and published
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes generate no new warnings
- [ ] New and existing tests pass with my changes
- [ ] I have made corresponding changes to the documentation if necessary

aerorahul committed:
add wdqms.py and wdqms_driver.py from https://github.com/kevindougherty-noaa/wdqms at hash f7728cc

aerorahul committed:
bring contents of wdqms_driver.py into wdqms.py.  Run wdqms.py through autopep8

aerorahul committed:
make wdqms.py into an executable

aerorahul committed:
add rocoto job, jjob, exjob and update configs for wdqms. also add it to rocoto

aerorahul committed:
add relevant ecf script and trigger it in the ecf suite def

KateFriedman-NOAA commented on parm/config/config.base.emc.dyn:
```diff
@@ -52,6 +52,7 @@ export DO_BUFRSND="YES"    # BUFR sounding products
 export DO_GEMPAK="NO"      # GEMPAK products
 export DO_AWIPS="NO"       # AWIPS products
 export WAFSF="NO"          # WAFS products
+export DO_WDWMS="NO"       # Generate WDQMS products
```
Should this be `DO_WDQMS` with a Q instead of that second W?

KateFriedman-NOAA commented on parm/config/config.base.nco.static:
```diff
@@ -50,6 +50,7 @@ export DO_BUFRSND="YES"    # BUFR sounding products
 export DO_GEMPAK="YES"     # GEMPAK products
 export DO_AWIPS="YES"      # AWIPS products
 export WAFSF="YES"         # WAFS products
+export DO_WDWMS="YES"      # Generate WDQMS products
```
Should this be DO_WDQMS with a Q instead of that second W?

KateFriedman-NOAA commented on parm/config/config.resources.emc.dyn:
```diff
@@ -176,6 +176,13 @@ elif [ $step = "analdiag" ]; then
 
+elif [ $step = "wdqms" ]; then
+
+    export wtime_wdqms="00:30:00"
+    export npe_wdqms=1
+    export nth_wdqms=1
+    export npe_node_wdqms=1
```
1. The ecf script has 10 minutes for the walltime but this has 30 minutes. Please correct one or the other.
2. Add `export memory_wdqms="48GB"`; otherwise the rocoto xml won't get the shared flag because it doesn't see a memory request.

KateFriedman-NOAA replied:
Thanks for making adjustments!

KateFriedman-NOAA commented on parm/config/config.resources.nco.static:
```diff
@@ -144,6 +144,13 @@ elif [ $step = "analdiag" ]; then
 
+elif [ $step = "wdqms" ]; then
+
+    export wtime_wdqms="00:30:00"
+    export npe_wdqms=1
+    export nth_wdqms=1
+    export npe_node_wdqms=1
```
Same comment as what I left for `config.resources.emc.dyn`.

KateFriedman-NOAA replied:
Thanks for making adjustments!

KateFriedman-NOAA commented on jobs/rocoto/wdqms.sh:
```diff
@@ -0,0 +1,14 @@
+#!/bin/bash
```
So all of the other rocoto job scripts in `dev/gfs.v16` are still ksh:
```
#!/bin/ksh -x
```
As much as I'd prefer the whole v16 package was converted to bash...perhaps for consistency make this ksh like the others? We do have a mix of shells across the v16 system though, so I won't push hard for this. Wanted to at least note the difference.

aerorahul replied:
ok. will update momentarily.

KateFriedman-NOAA commented on jobs/JGDAS_ATMOS_ANALYSIS_WDQMS:
```diff
@@ -0,0 +1,71 @@
+##############################################
+# Define COM directories
+##############################################
+export COMIN="Where are the files in INPUT_LIST in COM?"
+export COMOUT="Where should we place the wdqms.py generated files in COM"
```
Dropping comment here since it's TODO. Want to check on this later.

KateFriedman-NOAA commented on jobs/JGDAS_ATMOS_ANALYSIS_WDQMS:
```diff
@@ -0,0 +1,71 @@
+done
+
+###############################################################
+source "${BASE_ENV}/${machine}.env" wdqms
```
Do we need an entry for this job in the env file(s)?

aerorahul replied:
Actually we don't need since there is no `APRUN` command being set.  All this is doing is setting the machine specific environment variables at the top of the `MACHINE.env` file.  An argument is required since it checks for it and aborts if not provided.

KateFriedman-NOAA replied:
Gotcha, thanks!

kevindougherty-noaa reviewed (commented):
The python end looks good to me. I have heard there are new features that may need to be added in the future, is that something we should focus on for v17?

KateFriedman-NOAA commented on ecf/scripts/gdas/atmos/analysis/jgdas_atmos_analysis_wdqms.ecf:
```diff
@@ -0,0 +1,63 @@
+
+%include <tail.h>
+%manual
+######################################################################
+# Purpose: To execute the job that creates the WDQMS products
+######################################################################
+
+######################################################################
+# Job specific troubleshooting instructions:
+#  see generic troubleshoot manual page
+#
+######################################################################
+
+# include manual page below
```
This is new. Is this something we should be adding to all ecf scripts?

aerorahul replied:
yes. providing tooltip information in the ecflow UI on what the job is, is a good thing.

KateFriedman-NOAA commented on ecf/scripts/gdas/atmos/analysis/jgdas_atmos_analysis_wdqms.ecf:
```diff
@@ -0,0 +1,63 @@
+module load cray-pals/${cray_pals_ver}
+module load hdf5/${hdf5_ver}
+module load netcdf/${netcdf_ver}
+# TODO: Needs the python modules. Review the above list (hdf5 and netcdf)
```
Leaving comment to check on this later since it's TODO.

The module list looks good and matches the base set of modules that most jobs currently load in ops on WCOSS2. Other jobs load `python/3.8.6` set via `versions/run.ver`.

aerorahul replied:
sounds good.  I will load the python module.  We will need to ensure it contains pandas and whatever else the `wdqms.py` requires.

KateFriedman-NOAA reviewed (commented):
Can't comment on what the new job is actually doing, will lean on other reviewers for feedback on that aspect. Was able to review all other changes, including infrastructure for adding new job. Left a few comments to consider.

emilyhcliu commented:
Kevin,

To meet WDQMS requirements, we need to separate the BUOY and SHIP data from
the Marine file
To do so, both GSI and the WDQMS python code need to be modified. I have
the GSI and WDQMS branches for the changes.  I will issue PRs for you to
review.  These changes will be for V17.

Rahul PR #2372 <http://#2372> is for v16, no GSI code changes involved.

Emily


*Emily Huichun Liu*
Physical Scientist
Modeling and Data Assimilation Branch - DA & QC
NOAA/NWS/NCEP Environmental Modeling Center
National Centers for Weather and Climate Prediction (NCWCP)
5830 University Research Court
College Park, MD 20740
Phone: 301-683-3639
Cell: 410-948-6256



On Mon, Mar 4, 2024 at 10:34 AM kevindougherty-noaa <
***@***.***> wrote:

> ***@***.**** commented on this pull request.
>
> The python end looks good to me. I have heard there are new features that
> may need to be added in the future, is that something we should focus on
> for v17?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/global-workflow/pull/2372#pullrequestreview-1914635002>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AITLO5W2GIKKC3HMMUWP5E3YWSH7JAVCNFSM6AAAAABEFKWCGSVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTSMJUGYZTKMBQGI>
> .
> You are receiving this because your review was requested.Message ID:
> ***@***.***>
>

aerorahul committed:
address comments from review

aerorahul committed:
update paths to COMIN/COMOUT and unravel the cnvstat to get inputs for wdqms.py

aerorahul commented:
@emilyhcliu
Have you had a chance to test this branch out?

emilyhcliu commented:
@aerorahul I have set up the global workflow to test WDQMS on WCOSS2 but have not tested it yet. I was working on testing the new VIIRS NSST data from NESDIS for the next ObsProc upgrade (the due date is this week). I just completed testing the VIIRS NSST data this morning. My next task is to test the WDQMS on my list. I will complete the work by the end of tomorrow.

KateFriedman-NOAA committed:
Version update and new release notes for WDQMS

- Create skeleton release notes for WDQMS update in ops.
- Update system version to tentative v16.3.14. Confirm later.

Refs #2389

aerorahul committed:
Merge branch 'release/gfs.v16.3_wdqms' into feature/v16.wdqms

aerorahul commented:
closing this PR and opening another one directed towards the release branch.  
The feature branch is the same for testing etc.
